effectful/handlers/llm/providers.py

```diff
@@ -92,36 +94,79 @@ def _(values: Sequence) -> OpenAIMessageContent:
         return [{"type": "text", "text": str(values)}]
 
 
+@format_value.register(types.FunctionType)  # type: ignore
+def _(obj: types.FunctionType) -> OpenAIMessageContent:
+    try:
+        source = textwrap.dedent(inspect.getsource(obj)).strip()
+    except (OSError, TypeError):
+        doc = obj.__doc__ or "No docstring"
+        source = f"# <function {obj.__name__}>\n# {doc}"
+    return [{"type": "text", "text": source}]
+
+
+@format_value.register(type)  # type: ignore
+def _(obj: type) -> OpenAIMessageContent:
+    try:
+        source = textwrap.dedent(inspect.getsource(obj)).strip()
+    except (OSError, TypeError):
+        doc = obj.__doc__ or "No docstring"
+        source = f"# <class {obj.__name__}>\n# {doc}"
+    return [{"type": "text", "text": source}]
+
+
+@format_value.register(types.ModuleType)  # type: ignore
```

Comment:
nit: omit `types.ModuleType` argument, since it should be extractable from the signature